### PR TITLE
Feature: Manual Release Names

### PR DESCRIPTION
**Pull request by samblackk**
Attached to https://github.com/astronomer/issues/issues/548 

**This is a breaking change** 
Will need to ensure UI and CLI changes are merged in this together.

PR for the UI updates: https://github.com/astronomer/astro-ui/pull/507

This PR adds the required field, `releaseName` to the `createDeployment` mutation. The system already handles checking the releaseName against the database for uniqueness. 

Here's the console log after creating a deployment with the name space: `this-is-it` 
```
2020-04-22T13:29:30 INFO Final helm values: 
rbacEnabled: true
ingress:
    enabled: true
    baseDomain: local.astronomer.io
    class: deadly-terrier-nginx
    webserverAnnotations:
        nginx.ingress.kubernetes.io/server-snippet: "\n             location ^~ /api/ {\n               if ($host = 'deployments.local.astronomer.io' ) {\n                 return 404;\n               }\n               proxy_pass http://deployments.local.astronomer.io/airflow/$request_uri;\n             }"
        nginx.ingress.kubernetes.io/configuration-snippet: "\n             if ($host = 'deadly-terrier-airflow.local.astronomer.io' ) {\n               return 308 http://deployments.local.astronomer.io/airflow/$request_uri;\n             }"
        kubernetes.io/ingress.class: deadly-terrier-nginx
        nginx.ingress.kubernetes.io/custom-http-errors: '403,404,502,503'
        nginx.ingress.kubernetes.io/auth-url: 'http://houston.local.astronomer.io:8871/v1/authorization'
        nginx.ingress.kubernetes.io/auth-signin: 'http://app.local.astronomer.io:5000/login'
        nginx.ingress.kubernetes.io/proxy-cookie-path: '/ /this-is-it/'
        nginx.ingress.kubernetes.io/auth-response-headers: 'authorization, username, email'
    flowerAnnotations:
        nginx.ingress.kubernetes.io/rewrite-target: /$2;
        nginx.ingress.kubernetes.io/configuration-snippet: "\n             if ($host = 'deadly-terrier-flower.local.astronomer.io' ) {\n               rewrite ^ http://deployments.local.astronomer.io/flower permanent;\n             }\n             subs_filter_types text/css text/xml text/css;\n             sub_filter '=\"/' '=\"/this-is-it/flower/';\n             sub_filter_last_modified on;\n             sub_filter_once off;"
        kubernetes.io/ingress.class: deadly-terrier-nginx
        nginx.ingress.kubernetes.io/custom-http-errors: '403,404,502,503'
        nginx.ingress.kubernetes.io/auth-url: 'http://houston.local.astronomer.io:8871/v1/authorization'
        nginx.ingress.kubernetes.io/auth-signin: 'http://app.local.astronomer.io:5000/login'
        nginx.ingress.kubernetes.io/proxy-cookie-path: '/ /this-is-it/'
        nginx.ingress.kubernetes.io/auth-response-headers: 'authorization, username, email'
networkPolicies:
    enabled: true
pgbouncer:
    enabled: true
    resources:
        requests:
            cpu: 200m
            memory: 768Mi
        limits:
            cpu: 200m
            memory: 768Mi
    metadataPoolSize: 14
    maxClientConn: 140
allowPodLaunching: true
data:
    metadataConnection:
        user: this_is_it_airflow
        pass: VLXYU3jPTysHvr0uKuP9gwkp8QdkOFSb
        host: localhost
        port: 5432
        db: this_is_it_airflow
    resultBackendConnection:
        user: this_is_it_celery
        password: postgres
        host: localhost
        port: 5432
        database: postgres
        pass: BYM654MBw2hd8ZnuwxTPbmTHzkv5RR6r
        db: this_is_it_airflow
registry:
    connection:
        pass: 56Cec0iLNO1IAVvz9NwpHdNGapcyHWkp
        user: this-is-it
        host: registry.local.astronomer.io
        email: admin@local.astronomer.io
elasticsearch:
    connection:
        pass: FJO2UlZDbZfWn6EOPOPAujC1v9ahavws
        user: this-is-it
        host: localhost
        port: 9200
    enabled: false
fernetKey: MUdGZUo0MFB2SjFwUmV4WVM1N1dDbGJvOFJ0cnBiY2Q=
scheduler:
    resources:
        requests:
            cpu: 500m
            memory: 1920Mi
        limits:
            cpu: 500m
            memory: 1920Mi
webserver:
    resources:
        requests:
            cpu: 500m
            memory: 1920Mi
        limits:
            cpu: 500m
            memory: 1920Mi
statsd:
    resources:
        requests:
            cpu: 200m
            memory: 768Mi
        limits:
            cpu: 200m
            memory: 768Mi
flower:
    resources:
        requests:
            cpu: 200m
            memory: 768Mi
        limits:
            cpu: 200m
            memory: 768Mi
redis:
    resources:
        requests:
            cpu: 200m
            memory: 768Mi
        limits:
            cpu: 200m
            memory: 768Mi
workers:
    resources:
        requests:
            cpu: 1000m
            memory: 3840Mi
        limits:
            cpu: 1000m
            memory: 3840Mi
    terminationGracePeriodSeconds: 600
    replicas: 1
    persistence:
        enabled: true
limits:
    -
        type: Pod
        max:
            cpu: 3500m
            memory: 13440Mi
    -
        type: Container
        default:
            cpu: 100m
            memory: 384Mi
        defaultRequest:
            cpu: 100m
            memory: 384Mi
quotas:
    pods: 14
    requests.cpu: 6200m
    requests.memory: 23808Mi
    limits.cpu: 6200m
    limits.memory: 23808Mi
defaultAirflowTag: 0.0.0-alpine3.10-onbuild
labels:
    platform: deadly-terrier
    workspace: ck98vy4q3001k078160i3chaz
platform:
    release: deadly-terrier
    workspace: ck98vy4q3001k078160i3chaz
executor: CeleryExecutor
```

There may be more to handle here - I'm just not sure if/where.

### CODE REVIEW

File: src/resolvers/mutation/create-deployment/index.js [hunk 1]
```diff
@@ -95,7 +95,7 @@ export default async function createDeployment(parent, args, ctx, info) {
   ).toString("base64");
 
   // Generate a random space-themed release name.
-  const releaseName = generateReleaseName();
+  const releaseName = args.releaseName;
```
Comment (andriisoldatenko) [inline src/resolvers/mutation/create-deployment/index.js:98]: @samblackk wdyt about `args.releaseName || generateReleaseName();`

File: src/resolvers/mutation/create-deployment/index.js [hunk 2]
```diff
@@ -95,7 +95,7 @@ export default async function createDeployment(parent, args, ctx, info) {
   ).toString("base64");
 
   // Generate a random space-themed release name.
```
Comment (andriisoldatenko) [inline src/resolvers/mutation/create-deployment/index.js:97]: comment != implementation.

File: src/resolvers/mutation/create-deployment/index.js (hunk 1 above)
Comment (andriisoldatenko) [reply src/resolvers/mutation/create-deployment/index.js:98]: @schnie wdyt if we can add validation:
```
pattern = /^(?![0-9]+$)(?!.*-$)(?!-)[a-zA-Z0-9-]{1,63}$/g;
```
we can add tests :)

Comment (andriisoldatenko) [reply src/resolvers/mutation/create-deployment/index.js:98]: > I think we'd need to move the generation script client-side so the user can see it before sending it.

also i need to implement same in astro-cli too..

File: src/schema.graphql [hunk 3]
```diff
@@ -427,6 +427,7 @@ type Mutation {
 
   createDeployment(
     workspaceUuid: Uuid!
+    releaseName: String!
```
Comment (andriisoldatenko) [inline src/schema.graphql:430]: @schnie i think better for backward compatibility to make optional? and `default = args.releaseName || generateReleaseName();`

File: src/resolvers/mutation/create-deployment/index.js (hunk 1 above)
Comment (samblackk) [reply src/resolvers/mutation/create-deployment/index.js:98]: Perhaps we can make a small query that sends back a generated releaseName? That way can ensure the same script is being used and if we need to make changes - we're only changing 1 source.

Comment (andriisoldatenko) [reply src/resolvers/mutation/create-deployment/index.js:98]: sgtm!

**Comment by samblackk** [comment]
@andriisoldatenko Updated to include the new `appConfig` value: `manualReleaseNames` so we can flag this client-side and handle appropriately. I also updated the schema so `releaseName` isn't required and in `createDeployment` the script will either use the manual name if it's allowed and it's sent in the payload OR fall back to randomly generated.

File: config/default.yaml [hunk 4]
```diff
@@ -364,6 +364,9 @@ deployments:
   # Labels specifed here are applied to Airflow deployment namespaces.
   namespaceLabels: {}
 
+  # Enable manual namespace labels 
+  manualNamespaceLabels: false
```
Comment (samblackk) [inline config/default.yaml:368]: Defaults to false (cloud).

**Comment by samblackk** [comment]
@schnie Can I get a final review on this?

File: config/default.yaml (hunk 4 above)
Comment (schnie) [reply config/default.yaml:368]: @samblackk what do you think about `manualReleaseNames` here to match the resolver? I think it'd me slightly more accurate too.

File: src/resolvers/mutation/create-deployment/index.js [hunk 5]
```diff
@@ -94,8 +95,12 @@ export default async function createDeployment(parent, args, ctx, info) {
     generatePassword({ length: 32, numbers: true })
   ).toString("base64");
 
-  // Generate a random space-themed release name.
-  const releaseName = generateReleaseName();
+  // Set release name - either randomly generated or manually
+  const isManualNamespace =
+    config.get("deployments.manualNamespaceLabels") && args.releaseName;
```
Comment (schnie) [inline src/resolvers/mutation/create-deployment/index.js:100]: @samblackk this config needs to be updated I think

**Comment by samblackk** [comment]
Let's work on getting this merge today

**Comment by andriisoldatenko** [comment]
@samblackk nice work!
practically LGTM!
```
helm list -ANAME                    NAMESPACE                       REVISION        UPDATED                                 STATUS        CHART                            APP VERSION
andrii-awesomeasd-2     astronomer-andrii-awesomeasd-2  1               2020-05-05 23:30:20.901813433 +0000 UTC deployed      airflow-0.13.0-alpha.2
astronomer              astronomer                      6               2020-05-06 02:29:05.001623 +0300 EEST   deployed      astronomer-0.13.1-alpha.2        0.13.1-alpha.2
(
```

houston-logs:
```
2020-05-05T23:30:20 DEBUG ::ffff:10.244.0.1 - GET /v1/healthz HTTP/1.1 200 2 - 0.576 ms
2020-05-05T23:30:22 INFO Response from #createDeployment: {"result":{"success":true,"message":"Deployment Created"},"deployment":{"releaseName":"andrii-awesomeasd-2"}}
```

```

NAME                                             READY   STATUS    RESTARTS   AGE
andrii-awesomeasd-2-flower-5f99bf9b77-hhngn      1/1     Running   0          3m35s
andrii-awesomeasd-2-pgbouncer-55888b49b7-bxnzb   2/2     Running   0          3m35s
andrii-awesomeasd-2-redis-0                      1/1     Running   0          3m35s
andrii-awesomeasd-2-scheduler-8655d48695-k62sp   2/2     Running   0          3m35s
andrii-awesomeasd-2-statsd-df957df9b-cxdfz       1/1     Running   0          3m35s
andrii-awesomeasd-2-webserver-bcf6b4ff9-7jwkx    1/1     Running   0          3m35s
andrii-awesomeasd-2-worker-c7bf7474d-62gxf       1/1     Running   0          3m35s


```

**Comment by andriisoldatenko** [comment]
@schnie i think we are good to merge it together with https://github.com/astronomer/astro-cli/pull/327 and https://github.com/astronomer/astro-ui/pull/507

File: src/lib/deployments/naming/index.js [hunk 6]
```diff
@@ -1,9 +1,32 @@
 import adjectives from "./adjectives";
```
Comment (schnie) [inline src/lib/deployments/naming/index.js:1]: Minor thing, but I think we also have a `src/lib/deployments/validate` directory. Would this fit over there?

Comment (samblackk) [reply src/lib/deployments/naming/index.js:1]: It pertains to the actual name generating functions also in `deployments/naming`. It makes more sense to leave it where it is.

**Pull request merged**